### PR TITLE
Trainer callbacks

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR does two things: clean up a bit the files supporting `Trainer` and the `Trainer` class, and add callbacks to `Trainer`.

### Callbacks

This PR introduces a new class called `TrainerCallback` that can access the current state of the training loop and make some decisions (shown in the `TrainerControl` object). This allows us to isolate the pieces of code that do log-reporting on the various ML platforms or report progress in another file and clean up the code of the main `train` method of the `Trainer`.

This way, any new platform we want to integrate with for log-reporting or new behavior (like early stopping) can be implemented in a Callback while `Trainer` focuses on the main aspects of actual training, with or without mixed precision, on one or several GPUs/TPUs.

As an example, integrations with TensorBoard, Wandb and ComeML are moved to the `integrations` module in clean callbacks, while the control flow of logs/saves/evaluations as well as progress reporting are moved to the `trainer_callback` file.

Most of the behavior stays the same as this PR essentially moves code around, but there are a few API changes:
- deprecating the `tb_writer` argument in `Trainer` (with full backward compatibility), people should now use the `TensorBoardCallback`.
- a new `callbacks` argument in the `Trainer` init and new `add_callback`, `pop_callback` and `remove_callback` for the `Trainer`. For all of those, you can either pass an instance of a callback or a callback class.
- Cleaned up the progress bars a bit with only one main progress bar over all the steps we will do for training and evaluation bars that disappear after being done

### Progress bars

Here is the new progress bar behavior in console mode (checked in single and multi GPU envs, to make sure only one progress bar is displayed/logs are only printed once):

and in a jupyter notebook:

### General cleanup

Not directly related to this PR, but related to the general cleanup of `Trainer`, I moved a bit of stuff around: moved the utils at the start of `Trainer` to a new `trainer_utils_pt`. This way `trainer_utils` can be about the general training utils that work on both PyTorch and TensorFlow, and I moved the ones specific to PyTorch to `trainer_utils_pt`.

Also in `Trainer`, the code for logs, save and evaluation ended being duplicated between the end of a training step and the end of an epoch, so I put it in its private method to improve readability.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -465,102 +467,12 @@ def create_optimizer_and_scheduler(self, num_training_steps: int):
                 self.optimizer, num_warmup_steps=self.args.warmup_steps, num_training_steps=num_training_steps
             )
 
-    def setup_wandb(self):
```
Comment (sgugger) [inline src/transformers/trainer.py:468]: Removing this method is the only real breaking change, but I doubt we have a lot of users that subclassed Trainer to implement a custom `setup_wandb`. For those users, the new way is to subclass `WandbCallback` and override the `setup` method.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -723,17 +631,25 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
...
         self.state.num_train_epochs = num_train_epochs
+        self.state.is_local_process_zero = self.is_local_process_zero()
+        self.state.is_world_process_zero = self.is_world_process_zero()
 
         tr_loss = torch.tensor(0.0).to(self.args.device)
+        self._logging_loss_scalar = 0
         self._total_flos = self.state.total_flos
-        logging_loss_scalar = 0.0
         model.zero_grad()
-        disable_tqdm = self.args.disable_tqdm or not self.is_local_process_zero()
```
Comment (sgugger) [inline src/transformers/trainer.py:735]: Not commenting on all of those, but all references to progress bars disappear as it's all dealt with in ProgressCalback.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -863,9 +744,36 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
                 state_dict = torch.load(os.path.join(self.state.best_model_checkpoint, WEIGHTS_NAME))
                 self.model.load_state_dict(state_dict)
 
+        self.control = self.callback_handler.on_train_end(self.args, self.state, self.control)
+
         return TrainOutput(self.state.global_step, tr_loss.item() / self.state.global_step)
 
-    def _save_training(self, model, trial, metrics=None):
```
Comment (sgugger) [inline src/transformers/trainer.py:868]: Renamed private `_save_training` to `_save_checkpoint` as I think it's better.

File: docs/source/internal/trainer_utils.rst [hunk 4]
```diff
@@ -0,0 +1,21 @@
+Utilities for Trainer
+-----------------------------------------------------------------------------------------------------------------------
+
+This page lists all the utility functions used by :class:`~transformers.Trainer`.
+
+Most of those are only useful if you are studying the code of the tokenizers in the library.
```
Comment (LysandreJik) [inline docs/source/internal/trainer_utils.rst:6]:
```suggestion
Most of those are only useful if you are studying the code of the Trainer in the library.
```

File: src/transformers/integrations.py [hunk 5]
```diff
@@ -182,5 +179,159 @@ def _objective(trial, checkpoint_dir=None):
...
+
+
+class CometCallback(TrainerCallback):
+    """
+    A :class:`~transformers.TrainerCallback` that sends the logs to `Comet ML
+    <https://www.comet.ml/site/>`__.
+    """
+
+    def __init__(self):
+        assert _has_comet, "CometCallback requires wandb to be installed. Run `pip install comet-ml`."
```
Comment (LysandreJik) [inline src/transformers/integrations.py:291]: I don't think comet requires wandb to be installed :)

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -102,91 +131,21 @@
...
-        model (:class:`~transformers.PreTrainedModel`, `optional`):
+        model (:class:`~transformers.PreTrainedModel` or :obj:`torch.nn.Module`, `optional`):
             The model to train, evaluate or use for predictions. If not provided, a ``model_init`` must be passed.
+
+            .. note::
+
+                :class:`~transformers.Trainer` is optimized to work with the :class:`~transformers.PreTrainedModel`
+                provided by the library. You can still use your own models defined as :obj:`torch.nn.Module` as long as
+                they work the same way as the 🤗 Transformers models (expect input and labels as arguments and return
+                a tuple with the loss first if labels are provided).
```
Comment (LysandreJik) [inline src/transformers/trainer.py:148]: Imo it would be better to clearly define the API using Python typing, rather than rapidly mentioning what a model should expect and return. Something like:

```
[...]
You can still use your own models defined as :obj:`torch.nn.Module` as long as they work the same way as the 🤗 Transformers models: they should expect `input` and `labels` and return a tuple with the loss as the first output if the `labels` are provided.

class Model(nn.Module):
    def __call__(self, inputs: torch.Tensor, labels: torch.Tensor) -> Tuple[torch.Tensor]:
        # do something with the inputs and labels
        return loss
```

**Review by LysandreJik — APPROVED** [review]
This is very cool, I think it cleans up the Trainer class a lot. LGTM!

As discussed with @lhoestq and @julien-c, it would be nice to add in the documentation that these are Keas-style callbacks. When we hear callbacks, we might understand function callbacks, which are very different in practice.

EDIT: After discussing it with Sylvain, the point made above doesn't need to be applied as it isn't as ambiguous as I thought it would be. Citing Sylvain: it's callback as in events, similarly to what is done in VisualBasic or Delphi: on_click/on_touch etc.

File: src/transformers/integrations.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/integrations.py:291]: Thanks, good catch!

File: src/transformers/trainer.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/trainer.py:148]: This would fail since it needs to return loss, logits (or just logits, if no labels are provided). Maybe just leave it as "as long as they work the same way as the 🤗 Transformers models"?

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -863,9 +753,36 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
                 state_dict = torch.load(os.path.join(self.state.best_model_checkpoint, WEIGHTS_NAME))
                 self.model.load_state_dict(state_dict)
 
+        self.control = self.callback_handler.on_train_end(self.args, self.state, self.control)
+
         return TrainOutput(self.state.global_step, tr_loss.item() / self.state.global_step)
 
-    def _save_training(self, model, trial, metrics=None):
+    def _maybe_log_save_evalute(self, tr_loss, model, trial, epoch):
```
Comment (patrickvonplaten) [inline src/transformers/trainer.py:760]: (nit) Think it's better to have the `if self.control.should_log:` statement before the two calls of `_maybe_log_save_evaluate()` and rename the function to `_log_save_evaluate()` => "maybe" in a function name sounds a bit scary to the reader IMO.

File: src/transformers/trainer_callback.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+        self.callbacks = []
+        for cb in callbacks:
+            self.add_callback(cb)
+        self.model = model
+        self.optimizer = optimizer
+        self.lr_scheduler = lr_scheduler
+        self.train_dataloader = None
+        self.eval_dataloader = None
+
+        has_flow_callback = False
```
Comment (patrickvonplaten) [inline src/transformers/trainer_callback.py:280]: Could probably be simplified to `if not any(isinstance(cb, DefaultFlowCallback) for cb in self.callbacks): logger.warn(...)`

**Review by patrickvonplaten — APPROVED** [review]
Looks super clean to me! I don't know enough about the Trainer for an in-detail PR though...

**Comment by marrrcin** [comment]
Finally! Can't wait for this PR to be merged.
I've briefly looked at the code and from my understanding it should support this case, but correct me if I'm wrong:

_When saving checkpoints (including model weights as well as scheduler and optimizer states), I will be able to attach to this process and store the checkpoint in some external repository (i.e GCS / W&B artifact)_,

right?

**Comment by sgugger** [comment]
Yes, you will be able to inject custom behavior to the saved checkpoint with the `on_save` event.

File: src/transformers/trainer.py (hunk 1 above)
Comment (dsblank) [reply src/transformers/trainer.py:468]: Two breaking changes: `setup_comet()` was removed as well. But that is okay with me too. Just wanted to point that out if this is listed in a set of "breaking changes" in the release notes.
